[
    {
        "content": "<div class=\"codehilite\"><pre><span></span><code><span class=\"n\">C</span><span class=\"err\">:</span><span class=\"p\">\\</span><span class=\"n\">some-repo</span><span class=\"p\">&gt;</span> <span class=\"n\">leanproject</span> <span class=\"n\">up</span>\n<span class=\"no\">[WinError 5]</span> <span class=\"n\">Access</span> <span class=\"n\">is</span> <span class=\"n\">denied</span><span class=\"err\">:</span> <span class=\"s1\">&#39;C:\\\\some-repo\\\\_target\\\\deps\\\\mathlib\\\\.git\\\\objects\\\\82\\\\8911039fff7d6ef90ccd33eac0dbe03764acc8&#39;</span>\n</code></pre></div>",
        "id": 195840547,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1588254383
    },
    {
        "content": "<p>Maybe try <code>leanproject down</code>? /s</p>",
        "id": 195840641,
        "sender_full_name": "Johan Commelin",
        "timestamp": 1588254435
    },
    {
        "content": "<p>Do you have read/write access to that file?</p>",
        "id": 195840706,
        "sender_full_name": "Johan Commelin",
        "timestamp": 1588254463
    },
    {
        "content": "<p>As in, can you verify this using other means?</p>",
        "id": 195840827,
        "sender_full_name": "Johan Commelin",
        "timestamp": 1588254500
    },
    {
        "content": "<p>well I can open it in NotePad++</p>",
        "id": 195840979,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1588254569
    },
    {
        "content": "<p>and I can <code>cat</code> it in the same terminal</p>",
        "id": 195841108,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1588254614
    },
    {
        "content": "<p>Weird...</p>",
        "id": 195841407,
        "sender_full_name": "Johan Commelin",
        "timestamp": 1588254734
    },
    {
        "content": "<p>Can you edit it or is it read only?</p>",
        "id": 195841452,
        "sender_full_name": "Kevin Buzzard",
        "timestamp": 1588254762
    },
    {
        "content": "<p>open command prompt as administrator</p>",
        "id": 195841506,
        "sender_full_name": "Steffan",
        "timestamp": 1588254790
    },
    {
        "content": "<p>then try again?</p>",
        "id": 195841585,
        "sender_full_name": "Steffan",
        "timestamp": 1588254829
    },
    {
        "content": "<p>delete .git folder and try again, maybe?</p>",
        "id": 195841740,
        "sender_full_name": "Steffan",
        "timestamp": 1588254880
    },
    {
        "content": "<p>The risk with that is that you lose all your local branches</p>",
        "id": 195842074,
        "sender_full_name": "Kevin Buzzard",
        "timestamp": 1588255037
    },
    {
        "content": "<p>When I've seen errors like that before it's been stuff like disk corruption</p>",
        "id": 195842127,
        "sender_full_name": "Kevin Buzzard",
        "timestamp": 1588255063
    },
    {
        "content": "<p>Yes, but if you don't care about git so much, like me</p>",
        "id": 195842145,
        "sender_full_name": "Steffan",
        "timestamp": 1588255070
    },
    {
        "content": "<p>very interestingly, renaming it to <code>.git_old</code> doesn't help, but moving it to another place made it work</p>",
        "id": 195842229,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1588255089
    },
    {
        "content": "<p>it was very strange: after I rename it to <code>.git_old</code> it gives me the same error message with <code>.git</code> replaced with <code>.git_old</code></p>",
        "id": 195842260,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1588255109
    },
    {
        "content": "<p>what if you take the <code>.</code> off?</p>",
        "id": 195842290,
        "sender_full_name": "Steffan",
        "timestamp": 1588255126
    },
    {
        "content": "<p>When I've seen copying stuff to other places fixing things, it has been disk corruption</p>",
        "id": 195842333,
        "sender_full_name": "Kevin Buzzard",
        "timestamp": 1588255144
    },
    {
        "content": "<p>I have reproduced this error</p>",
        "id": 195946935,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1588327564
    },
    {
        "content": "<ol>\n<li>delete the .git folder</li>\n<li><code>leanproject up</code></li>\n<li>this works fine, and it clones mathlib</li>\n<li><code>leanproject up</code></li>\n<li>error</li>\n</ol>",
        "id": 195946954,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1588327588
    },
    {
        "content": "<p>so basically it cannot be executed twice</p>",
        "id": 195946965,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1588327604
    },
    {
        "content": "<p>Open an issue in the GitHub repo and Patrick will look at it when he has time. Nice sleuthing.</p>",
        "id": 195947076,
        "sender_full_name": "Kevin Buzzard",
        "timestamp": 1588327729
    },
    {
        "content": "<p>the environment is in the VSCode Terminal on Windows 10</p>",
        "id": 195947083,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1588327742
    },
    {
        "content": "<p>which uses powershell</p>",
        "id": 195947091,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1588327753
    },
    {
        "content": "<p>debug:</p>\n<div class=\"codehilite\"><pre><span></span><code>Traceback (most recent call last):\n  File &quot;c:\\python-3.6.4\\lib\\runpy.py&quot;, line 193, in _run_module_as_main\n    &quot;__main__&quot;, mod_spec)\n  File &quot;c:\\python-3.6.4\\lib\\runpy.py&quot;, line 85, in _run_code\n    exec(code, run_globals)\n  File &quot;C:\\Python-3.6.4\\Scripts\\leanproject.exe\\__main__.py&quot;, line 7, in &lt;module&gt;\n  File &quot;c:\\python-3.6.4\\lib\\site-packages\\mathlibtools\\leanproject.py&quot;, line 256, in safe_cli\n    handle_exception(err, str(err))\n  File &quot;c:\\python-3.6.4\\lib\\site-packages\\mathlibtools\\leanproject.py&quot;, line 51, in handle_exception\n    raise exc\n  File &quot;c:\\python-3.6.4\\lib\\site-packages\\mathlibtools\\leanproject.py&quot;, line 254, in safe_cli\n    cli()\n  File &quot;c:\\python-3.6.4\\lib\\site-packages\\click\\core.py&quot;, line 829, in __call__\n    return self.main(*args, **kwargs)\n  File &quot;c:\\python-3.6.4\\lib\\site-packages\\click\\core.py&quot;, line 782, in main\n    rv = self.invoke(ctx)\n  File &quot;c:\\python-3.6.4\\lib\\site-packages\\click\\core.py&quot;, line 1259, in invoke\n    return _process_result(sub_ctx.command.invoke(sub_ctx))\n  File &quot;c:\\python-3.6.4\\lib\\site-packages\\click\\core.py&quot;, line 1066, in invoke\n    return ctx.invoke(self.callback, **ctx.params)\n  File &quot;c:\\python-3.6.4\\lib\\site-packages\\click\\core.py&quot;, line 610, in invoke\n    return callback(*args, **kwargs)\n  File &quot;c:\\python-3.6.4\\lib\\site-packages\\mathlibtools\\leanproject.py&quot;, line 92, in upgrade_mathlib\n    proj().upgrade_mathlib()\n  File &quot;c:\\python-3.6.4\\lib\\site-packages\\mathlibtools\\lib.py&quot;, line 468, in upgrade_mathlib\n    shutil.rmtree(str(self.mathlib_folder))\n  File &quot;c:\\python-3.6.4\\lib\\shutil.py&quot;, line 494, in rmtree\n    return _rmtree_unsafe(path, onerror)\n  File &quot;c:\\python-3.6.4\\lib\\shutil.py&quot;, line 384, in _rmtree_unsafe\n    _rmtree_unsafe(fullname, onerror)\n  File &quot;c:\\python-3.6.4\\lib\\shutil.py&quot;, line 384, in _rmtree_unsafe\n    _rmtree_unsafe(fullname, onerror)\n  File &quot;c:\\python-3.6.4\\lib\\shutil.py&quot;, line 384, in _rmtree_unsafe\n    _rmtree_unsafe(fullname, onerror)\n  File &quot;c:\\python-3.6.4\\lib\\shutil.py&quot;, line 389, in _rmtree_unsafe\n    onerror(os.unlink, fullname, sys.exc_info())\n  File &quot;c:\\python-3.6.4\\lib\\shutil.py&quot;, line 387, in _rmtree_unsafe\n    os.unlink(fullname)\nPermissionError: [WinError 5] Access is denied: &#39;C:\\\\some-repo\\\\_target\\\\deps\\\\mathlib\\\\.git\\\\objects\\\\pack\\\\pack-36f591561703d876c8ad32e4cc46887e2a0bf6d2.idx&#39;\n</code></pre></div>",
        "id": 195947267,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1588327913
    },
    {
        "content": "<p><a href=\"https://github.com/leanprover-community/mathlib-tools/issues/46\" title=\"https://github.com/leanprover-community/mathlib-tools/issues/46\">https://github.com/leanprover-community/mathlib-tools/issues/46</a></p>",
        "id": 195947420,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1588328048
    },
    {
        "content": "<p>I tried to reproduce, but you need to give more details. Here is what I did, in a folder containing a cloned copy of mathlib: I deleted the <code>.git</code> folder, and then I ran <code>leanproject up</code> several times. The output is</p>\n<div class=\"codehilite\"><pre><span></span><code><span class=\"n\">PS</span> <span class=\"n\">C</span><span class=\"o\">:</span><span class=\"err\">\\</span><span class=\"n\">Users</span><span class=\"err\">\\</span><span class=\"n\">Sebastien</span><span class=\"err\">\\</span><span class=\"n\">Desktop</span><span class=\"err\">\\</span><span class=\"n\">mathlib_klau</span><span class=\"bp\">&gt;</span> <span class=\"n\">leanproject</span> <span class=\"n\">up</span>\n<span class=\"err\">&#39;</span><span class=\"n\">dependencies&#39;</span>\n<span class=\"n\">PS</span> <span class=\"n\">C</span><span class=\"o\">:</span><span class=\"err\">\\</span><span class=\"n\">Users</span><span class=\"err\">\\</span><span class=\"n\">Sebastien</span><span class=\"err\">\\</span><span class=\"n\">Desktop</span><span class=\"err\">\\</span><span class=\"n\">mathlib_klau</span><span class=\"bp\">&gt;</span> <span class=\"n\">leanproject</span> <span class=\"n\">up</span>\n<span class=\"err\">&#39;</span><span class=\"n\">dependencies&#39;</span>\n<span class=\"n\">PS</span> <span class=\"n\">C</span><span class=\"o\">:</span><span class=\"err\">\\</span><span class=\"n\">Users</span><span class=\"err\">\\</span><span class=\"n\">Sebastien</span><span class=\"err\">\\</span><span class=\"n\">Desktop</span><span class=\"err\">\\</span><span class=\"n\">mathlib_klau</span><span class=\"bp\">&gt;</span> <span class=\"n\">leanproject</span> <span class=\"n\">up</span>\n<span class=\"err\">&#39;</span><span class=\"n\">dependencies&#39;</span>\n</code></pre></div>\n\n\n<p>It doesn't create a <code>.git</code> folder. I am not surprised it doesn't work, because <code>leanproject up</code> is supposed to use <code>.git</code>.</p>",
        "id": 195947953,
        "sender_full_name": "Sebastien Gouezel",
        "timestamp": 1588328604
    },
    {
        "content": "<p>I should specify that this is in a repo using mathlib</p>",
        "id": 195948055,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1588328691
    },
    {
        "content": "<p>This is still not enough for me. Details that you did not specify: is mathlib in a different directory with its own git structure, or is it in a subdirectory of your project, what does the <code>.toml</code> look like, and so on. Ideally, I would just have a github branch to clone, and then I could start experimenting.</p>",
        "id": 195948353,
        "sender_full_name": "Sebastien Gouezel",
        "timestamp": 1588328958
    },
    {
        "content": "<p><span class=\"user-mention\" data-user-id=\"110050\">@Sebastien Gouezel</span> <a href=\"https://github.com/kckennylau/test\" title=\"https://github.com/kckennylau/test\">https://github.com/kckennylau/test</a></p>",
        "id": 195948996,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1588329538
    },
    {
        "content": "<p><span class=\"user-mention silent\" data-user-id=\"110064\">Kenny Lau</span> <a href=\"#narrow/stream/113488-general/topic/leanproject.20access.20denied/near/195948996\" title=\"#narrow/stream/113488-general/topic/leanproject.20access.20denied/near/195948996\">said</a>:</p>\n<blockquote>\n<p><span class=\"user-mention silent\" data-user-id=\"110050\">Sebastien Gouezel</span> <a href=\"https://github.com/kckennylau/test\" title=\"https://github.com/kckennylau/test\">https://github.com/kckennylau/test</a></p>\n</blockquote>\n<p>I cloned your repo. <code>leanproject up</code> feels a bit slow when i run it (compared to when I run in mathlib). But it succeeds. Same when I run the command a 2nd time.</p>",
        "id": 195950443,
        "sender_full_name": "Johan Commelin",
        "timestamp": 1588330708
    },
    {
        "content": "<p>FWIW I can reproduce this on Windows, both Powershell and Git Bash. Running Powershell as admin doesn't fix it, I don't immediately see an issue with the file permissions, I don't see a process blocking write access to that file. Out of time to troubleshoot further right now though.</p>",
        "id": 195951564,
        "sender_full_name": "Rob Lewis",
        "timestamp": 1588331828
    },
    {
        "content": "<p><span class=\"user-mention\" data-user-id=\"110596\">@Rob Lewis</span> thanks for your reproduction.</p>",
        "id": 195951886,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1588332108
    },
    {
        "content": "<p>I am doing something wrong:</p>\n<div class=\"codehilite\"><pre><span></span><code>PS C:\\Users\\Sebastien\\Desktop\\essai&gt; leanproject --version\nleanproject, version 0.0.5\nPS C:\\Users\\Sebastien\\Desktop\\essai&gt; git clone https://github.com/kckennylau/test.git\nCloning into &#39;test&#39;...\nremote: Enumerating objects: 6, done.\nremote: Counting objects: 100% (6/6), done.\nremote: Compressing objects: 100% (3/3), done.\nremote: Total 6 (delta 0), reused 6 (delta 0), pack-reused 0\nUnpacking objects: 100% (6/6), done.\nPS C:\\Users\\Sebastien\\Desktop\\essai&gt; cd test\nPS C:\\Users\\Sebastien\\Desktop\\essai\\test&gt; leanproject up\n[WinError 2] Le fichier spécifié est introuvable\n</code></pre></div>",
        "id": 195956132,
        "sender_full_name": "Sebastien Gouezel",
        "timestamp": 1588336247
    },
    {
        "content": "<p>This is really windows specific. I couldn't reproduce here.</p>",
        "id": 195956529,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1588336560
    },
    {
        "content": "<p>Did you try <code>leanproject --debug up</code>?</p>",
        "id": 195956574,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1588336571
    },
    {
        "content": "<p>Are you asking Kenny or me?</p>\n<p>For me, I get</p>\n<div class=\"codehilite\"><pre><span></span><code>PS C:\\Users\\Sebastien\\Desktop\\essai\\test&gt; leanproject --debug up\nTraceback (most recent call last):\n  File &quot;C:\\Program Files\\Python37\\Scripts\\leanproject-script.py&quot;, line 11, in &lt;module&gt;\n    load_entry_point(&#39;mathlibtools==0.0.5&#39;, &#39;console_scripts&#39;, &#39;leanproject&#39;)()\n  File &quot;c:\\program files\\python37\\lib\\site-packages\\mathlibtools\\leanproject.py&quot;, line 295, in safe_cli\n    handle_exception(err, str(err))\n  File &quot;c:\\program files\\python37\\lib\\site-packages\\mathlibtools\\leanproject.py&quot;, line 51, in handle_exception\n    raise exc\n  File &quot;c:\\program files\\python37\\lib\\site-packages\\mathlibtools\\leanproject.py&quot;, line 293, in safe_cli\n    cli()\n  File &quot;c:\\program files\\python37\\lib\\site-packages\\click\\core.py&quot;, line 764, in __call__\n    return self.main(*args, **kwargs)\n  File &quot;c:\\program files\\python37\\lib\\site-packages\\click\\core.py&quot;, line 717, in main\n    rv = self.invoke(ctx)\n  File &quot;c:\\program files\\python37\\lib\\site-packages\\click\\core.py&quot;, line 1137, in invoke\n    return _process_result(sub_ctx.command.invoke(sub_ctx))\n  File &quot;c:\\program files\\python37\\lib\\site-packages\\click\\core.py&quot;, line 956, in invoke\n    return ctx.invoke(self.callback, **ctx.params)\n  File &quot;c:\\program files\\python37\\lib\\site-packages\\click\\core.py&quot;, line 555, in invoke\n    return callback(*args, **kwargs)\n  File &quot;c:\\program files\\python37\\lib\\site-packages\\mathlibtools\\leanproject.py&quot;, line 93, in upgrade_mathlib\n    proj().upgrade_mathlib()\n  File &quot;c:\\program files\\python37\\lib\\site-packages\\mathlibtools\\lib.py&quot;, line 592, in upgrade_mathlib\n    self.run([&#39;leanpkg&#39;, &#39;upgrade&#39;])\n  File &quot;c:\\program files\\python37\\lib\\site-packages\\mathlibtools\\lib.py&quot;, line 528, in run\n    check=True).stdout.decode()\n  File &quot;c:\\program files\\python37\\lib\\subprocess.py&quot;, line 472, in run\n    with Popen(*popenargs, **kwargs) as process:\n  File &quot;c:\\program files\\python37\\lib\\subprocess.py&quot;, line 775, in __init__\n    restore_signals, start_new_session)\n  File &quot;c:\\program files\\python37\\lib\\subprocess.py&quot;, line 1178, in _execute_child\n    startupinfo)\nFileNotFoundError: [WinError 2] Le fichier spécifié est introuvable\n</code></pre></div>\n\n\n<p>But I want to stress that, for my usual case (<code>leanproject up</code> in a mathlib directory), everything works perfectly.</p>",
        "id": 195956697,
        "sender_full_name": "Sebastien Gouezel",
        "timestamp": 1588336695
    },
    {
        "content": "<p><span class=\"user-mention\" data-user-id=\"110031\">@Patrick Massot</span> I have produced the output of <code>leanproject --debug up</code> in the issue</p>",
        "id": 195960639,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1588339741
    },
    {
        "content": "<p>It fails to remove files in the mathlib folder. What happens if you try to remove it by hand?</p>",
        "id": 195965073,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1588342755
    },
    {
        "content": "<p>if I remove the <code>.git</code> folder then <code>leanproject up</code> does work, but only once</p>",
        "id": 195966413,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1588343542
    },
    {
        "content": "<p>I just ran into this in practice. The <code>.idx</code> file does indeed seem to be read-only, which prevents <code>shutil.rmtree</code> from deleting it.</p>",
        "id": 196605635,
        "sender_full_name": "Rob Lewis",
        "timestamp": 1588755731
    },
    {
        "content": "<p>One workaround: <a href=\"https://stackoverflow.com/questions/21261132/shutil-rmtree-to-remove-readonly-files\" title=\"https://stackoverflow.com/questions/21261132/shutil-rmtree-to-remove-readonly-files\">https://stackoverflow.com/questions/21261132/shutil-rmtree-to-remove-readonly-files</a> although blindly deleting read-only stuff is maybe not the smartest thing to do.</p>",
        "id": 196605673,
        "sender_full_name": "Rob Lewis",
        "timestamp": 1588755754
    },
    {
        "content": "<p>What is this .idx file?</p>",
        "id": 196605936,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1588755900
    },
    {
        "content": "<p>Oh, in git?</p>",
        "id": 196606021,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1588755944
    },
    {
        "content": "<p>Windows is so weird</p>",
        "id": 196606033,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1588755951
    },
    {
        "content": "<div class=\"codehilite\"><pre><span></span><code>robyl@DESKTOP-SMPP1GQ MINGW64 /d/lean/doc-gen <span class=\"o\">(</span>master<span class=\"o\">)</span>\n$ leanproject up\n<span class=\"o\">[</span>WinError <span class=\"m\">5</span><span class=\"o\">]</span> Access is denied: <span class=\"s1\">&#39;D:\\\\lean\\\\doc-gen\\\\_target\\\\deps\\\\mathlib\\\\.git\\\\objects\\\\pack\\\\pack-1be031bb56944e0cd6652e34e4ab97cadbe780f9.idx&#39;</span>\n</code></pre></div>",
        "id": 196606100,
        "sender_full_name": "Rob Lewis",
        "timestamp": 1588755965
    },
    {
        "content": "<p>If this is only .idx files in .git we can have a rule for them instead of blindly deleting read-only files</p>",
        "id": 196606191,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1588756014
    },
    {
        "content": "<p>Feel free to submit a patch to leanproject.</p>",
        "id": 196606206,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1588756027
    },
    {
        "content": "<p>workaround seems to be to execute <code>leanproject up</code> inside <code>_target/deps/mathlib</code></p>",
        "id": 198650415,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1590399182
    },
    {
        "content": "<p>push: <a href=\"https://github.com/leanprover-community/mathlib-tools/issues/46\">https://github.com/leanprover-community/mathlib-tools/issues/46</a></p>",
        "id": 198650422,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1590399191
    },
    {
        "content": "<p>If you do that then you should probably manually edit the <code>leanpkg.toml</code> in your repo so it correctly reports which mathlib commit it's using</p>",
        "id": 198652249,
        "sender_full_name": "Kevin Buzzard",
        "timestamp": 1590400677
    },
    {
        "content": "<p><span class=\"user-mention silent\" data-user-id=\"110064\">Kenny Lau</span> <a href=\"#narrow/stream/113488-general/topic/leanproject.20access.20denied/near/198650415\">said</a>:</p>\n<blockquote>\n<p>workaround seems to be to execute <code>leanproject up</code> inside <code>_target/deps/mathlib</code></p>\n</blockquote>\n<p>No, this is not at all a workaround.</p>",
        "id": 198664476,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1590410645
    },
    {
        "content": "<p>I blindly tried to push a workaround to the winbug branch, but I need Windows users to test it.</p>",
        "id": 198665823,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1590411571
    },
    {
        "content": "<p>(it's just a patch that installs linux)</p>",
        "id": 198666283,
        "sender_full_name": "Reid Barton",
        "timestamp": 1590411859
    },
    {
        "content": "<p><span class=\"user-mention\" data-user-id=\"110031\">@Patrick Massot</span> how do I test it?</p>",
        "id": 198666788,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1590412201
    },
    {
        "content": "<p>git clone the repo, <code>git checkout winbug</code>, <code>pip install .</code> and try to use leanproject as usual</p>",
        "id": 198667318,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1590412430
    },
    {
        "content": "<p>but how do I differentiate between the two different <code>leanproject</code>s?</p>",
        "id": 198667443,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1590412487
    },
    {
        "content": "<p>if I want to switch back to the original one do I <code>pip install mathlibtools</code>?</p>",
        "id": 198667459,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1590412498
    },
    {
        "content": "<p>yes</p>",
        "id": 198667624,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1590412578
    },
    {
        "content": "<p>pip installing the clone repo will overwrite your install, but you can switch back as you wrote</p>",
        "id": 198667653,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1590412600
    },
    {
        "content": "<p>Also, if you help enough then we can hopefully fix it today and it will become the official version</p>",
        "id": 198667704,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1590412635
    },
    {
        "content": "<div class=\"codehilite\"><pre><span></span><code>Traceback (most recent call last):\n  File &quot;C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.7_3.7.2032.0_x64__qbz5n2kfra8p0\\lib\\runpy.py&quot;, line 193, in _run_module_as_main\n    &quot;__main__&quot;, mod_spec)\n  File &quot;C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.7_3.7.2032.0_x64__qbz5n2kfra8p0\\lib\\runpy.py&quot;, line 85, in _run_code\n    exec(code, run_globals)\n  File &quot;C:\\Users\\Kenny Lau\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python37\\Scripts\\leanproject.exe\\__main__.py&quot;, line 7, in &lt;module&gt;\n  File &quot;C:\\Users\\Kenny Lau\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python37\\site-packages\\mathlibtools\\leanproject.py&quot;, line 308, in safe_cli\n    handle_exception(err, str(err))\n  File &quot;C:\\Users\\Kenny Lau\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python37\\site-packages\\mathlibtools\\leanproject.py&quot;, line 58, in handle_exception\n    raise exc\n  File &quot;C:\\Users\\Kenny Lau\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python37\\site-packages\\mathlibtools\\leanproject.py&quot;, line 306, in safe_cli\n    cli() # pylint: disable=no-value-for-parameter\n  File &quot;C:\\Users\\Kenny Lau\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python37\\site-packages\\click\\core.py&quot;, line 829, in __call__\n    return self.main(*args, **kwargs)\n  File &quot;C:\\Users\\Kenny Lau\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python37\\site-packages\\click\\core.py&quot;, line 782, in main\n    rv = self.invoke(ctx)\n  File &quot;C:\\Users\\Kenny Lau\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python37\\site-packages\\click\\core.py&quot;, line 1259, in invoke\n    return _process_result(sub_ctx.command.invoke(sub_ctx))\n  File &quot;C:\\Users\\Kenny Lau\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python37\\site-packages\\click\\core.py&quot;, line 1066, in invoke\n    return ctx.invoke(self.callback, **ctx.params)\n  File &quot;C:\\Users\\Kenny Lau\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python37\\site-packages\\click\\core.py&quot;, line 610, in invoke\n    return callback(*args, **kwargs)\n  File &quot;C:\\Users\\Kenny Lau\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python37\\site-packages\\mathlibtools\\leanproject.py&quot;, line 100, in upgrade_mathlib\n    proj().upgrade_mathlib()\n  File &quot;C:\\Users\\Kenny Lau\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python37\\site-packages\\mathlibtools\\lib.py&quot;, line 598, in upgrade_mathlib\n    shutil.rmtree(str(self.mathlib_folder))\n  File &quot;C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.7_3.7.2032.0_x64__qbz5n2kfra8p0\\lib\\shutil.py&quot;, line 516, in\nrmtree\n    return _rmtree_unsafe(path, onerror)\n  File &quot;C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.7_3.7.2032.0_x64__qbz5n2kfra8p0\\lib\\shutil.py&quot;, line 395, in\n_rmtree_unsafe\n    _rmtree_unsafe(fullname, onerror)\n  File &quot;C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.7_3.7.2032.0_x64__qbz5n2kfra8p0\\lib\\shutil.py&quot;, line 395, in\n_rmtree_unsafe\n    _rmtree_unsafe(fullname, onerror)\n  File &quot;C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.7_3.7.2032.0_x64__qbz5n2kfra8p0\\lib\\shutil.py&quot;, line 395, in\n_rmtree_unsafe\n    _rmtree_unsafe(fullname, onerror)\n  File &quot;C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.7_3.7.2032.0_x64__qbz5n2kfra8p0\\lib\\shutil.py&quot;, line 400, in\n_rmtree_unsafe\n    onerror(os.unlink, fullname, sys.exc_info())\n  File &quot;C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.7_3.7.2032.0_x64__qbz5n2kfra8p0\\lib\\shutil.py&quot;, line 398, in\n_rmtree_unsafe\n    os.unlink(fullname)\nPermissionError: [WinError 5] Access is denied: &#39;C:\\\\[repo]\\\\_target\\\\deps\\\\mathlib\\\\.git\\\\objects\\\\pack\\\\pack-49c3746f3db30550d71ee61c3234d10c6308216c.pack&#39;\n</code></pre></div>",
        "id": 198668798,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1590413310
    },
    {
        "content": "<p>Try to replace <code>*.idx</code> by <code>*</code> at <a href=\"https://github.com/leanprover-community/mathlib-tools/commit/6e06675add7a35228e2b81d9eb11a2e510ade538#diff-a97f9d481b2b8a9dffffdebc0e0de1d1R595\">https://github.com/leanprover-community/mathlib-tools/commit/6e06675add7a35228e2b81d9eb11a2e510ade538#diff-a97f9d481b2b8a9dffffdebc0e0de1d1R595</a></p>",
        "id": 198669033,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1590413457
    },
    {
        "content": "<p>also how do I tell pip to not install it in a deep directory?</p>",
        "id": 198669431,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1590413710
    },
    {
        "content": "<p>as in, not <code>C:\\Users\\Kenny Lau\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python37\\Scripts\\leanproject.exe</code></p>",
        "id": 198669444,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1590413722
    },
    {
        "content": "<p><code>pip install -e .</code></p>",
        "id": 198669450,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1590413724
    },
    {
        "content": "<p>but it shouldn't matter</p>",
        "id": 198669473,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1590413737
    },
    {
        "content": "<p>why do you need to access this install path at all?</p>",
        "id": 198669480,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1590413751
    },
    {
        "content": "<p>because <code>pip</code> told me that <code>pip</code> installed it there and it is not in PATH</p>",
        "id": 198669547,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1590413770
    },
    {
        "content": "<p>Then your system is broken</p>",
        "id": 198669661,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1590413858
    },
    {
        "content": "<div class=\"codehilite\"><pre><span></span><code>  File &quot;c:\\python-3.6.4\\lib\\runpy.py&quot;, line 193, in _run_module_as_main\n    &quot;__main__&quot;, mod_spec)\n  File &quot;c:\\python-3.6.4\\lib\\runpy.py&quot;, line 85, in _run_code\n    exec(code, run_globals)\n  File &quot;C:\\Python-3.6.4\\Scripts\\leanproject.exe\\__main__.py&quot;, line 7, in &lt;module&gt;\n  File &quot;c:\\python-3.6.4\\lib\\site-packages\\mathlibtools\\leanproject.py&quot;, line 256, in safe_cli\n    handle_exception(err, str(err))\n  File &quot;c:\\python-3.6.4\\lib\\site-packages\\mathlibtools\\leanproject.py&quot;, line 51, in handle_exception\n    raise exc\n  File &quot;c:\\python-3.6.4\\lib\\site-packages\\mathlibtools\\leanproject.py&quot;, line 254, in safe_cli\n    cli()\n  File &quot;c:\\python-3.6.4\\lib\\site-packages\\click\\core.py&quot;, line 829, in __call__\n    return self.main(*args, **kwargs)\n  File &quot;c:\\python-3.6.4\\lib\\site-packages\\click\\core.py&quot;, line 782, in main\n    rv = self.invoke(ctx)\n  File &quot;c:\\python-3.6.4\\lib\\site-packages\\click\\core.py&quot;, line 1259, in invoke\n    return _process_result(sub_ctx.command.invoke(sub_ctx))\n  File &quot;c:\\python-3.6.4\\lib\\site-packages\\click\\core.py&quot;, line 1066, in invoke\n    return ctx.invoke(self.callback, **ctx.params)\n  File &quot;c:\\python-3.6.4\\lib\\site-packages\\click\\core.py&quot;, line 610, in invoke\n    return callback(*args, **kwargs)\n  File &quot;c:\\python-3.6.4\\lib\\site-packages\\mathlibtools\\leanproject.py&quot;, line 92, in upgrade_mathlib\n    proj().upgrade_mathlib()\n  File &quot;c:\\python-3.6.4\\lib\\site-packages\\mathlibtools\\lib.py&quot;, line 468, in upgrade_mathlib\n    shutil.rmtree(str(self.mathlib_folder))\n  File &quot;c:\\python-3.6.4\\lib\\shutil.py&quot;, line 494, in rmtree\n    return _rmtree_unsafe(path, onerror)\n  File &quot;c:\\python-3.6.4\\lib\\shutil.py&quot;, line 384, in _rmtree_unsafe\n    _rmtree_unsafe(fullname, onerror)\n  File &quot;c:\\python-3.6.4\\lib\\shutil.py&quot;, line 384, in _rmtree_unsafe\n    _rmtree_unsafe(fullname, onerror)\n  File &quot;c:\\python-3.6.4\\lib\\shutil.py&quot;, line 384, in _rmtree_unsafe\n    _rmtree_unsafe(fullname, onerror)\n  File &quot;c:\\python-3.6.4\\lib\\shutil.py&quot;, line 389, in _rmtree_unsafe\n    onerror(os.unlink, fullname, sys.exc_info())\n  File &quot;c:\\python-3.6.4\\lib\\shutil.py&quot;, line 387, in _rmtree_unsafe\n    os.unlink(fullname)\nPermissionError: [WinError 5] Access is denied: &#39;C:\\\\[repo]\\\\_target\\\\deps\\\\mathlib\\\\.git\\\\objects\\\\pack\\\\pack-49c3746f3db30550d71ee61c3234d10c6308216c.pack&#39;\n</code></pre></div>",
        "id": 198669767,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1590413912
    },
    {
        "content": "<p><a href=\"https://www.pcmag.com/how-to/how-to-make-the-switch-from-windows-to-linux\">https://www.pcmag.com/how-to/how-to-make-the-switch-from-windows-to-linux</a></p>",
        "id": 198669771,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1590413918
    },
    {
        "content": "<p>Are you sure you modified that line?</p>",
        "id": 198669851,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1590413992
    },
    {
        "content": "<p>And rerun pip install if you didn't use <code>-e</code> in the first place</p>",
        "id": 198669916,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1590414014
    },
    {
        "content": "<p>pip doesn't tell me this time where it installed mathlibtools in</p>",
        "id": 198670012,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1590414094
    },
    {
        "content": "<p>so I assumed it's in PATH</p>",
        "id": 198670018,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1590414097
    },
    {
        "content": "<p>maybe I shouldn't use <code>-e</code> to make sure</p>",
        "id": 198670349,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1590414285
    },
    {
        "content": "<p>Honestly I don't understand how it can be that slow to iterate on this. But I'm not used to Windows hell</p>",
        "id": 198670440,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1590414359
    },
    {
        "content": "<p>hey it worked!</p>",
        "id": 198670510,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1590414391
    },
    {
        "content": "<p>Great</p>",
        "id": 198670755,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1590414562
    },
    {
        "content": "<p>I'll push this fix</p>",
        "id": 198670765,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1590414566
    },
    {
        "content": "<p>This is now live on pypi. You can update your official copy using <code>pip install --update mathlibtools</code></p>",
        "id": 198671302,
        "sender_full_name": "Patrick Massot",
        "timestamp": 1590414908
    },
    {
        "content": "<p>why does it keep installing in a remote path</p>",
        "id": 198671784,
        "sender_full_name": "Kenny Lau",
        "timestamp": 1590415241
    }
]